### PR TITLE
feat: implemented rbac module and added tests for roles

### PR DESCRIPTION
**Pull request by MrUnfunny**

### CODE REVIEW

File: lib/src/rbac/default_role_manager.dart [hunk 1]
```diff
@@ -12,46 +12,114 @@
...
+
+  void addMatchingFunc(dynamic name, MatchingFunc? func) {
+    hasPattern = true;
+
+    if (name.runtimeType == String || func != null) {
+      matchingFunc = func;
+    } else if (name.runtimeType == MatchingFunc) {
+      matchingFunc = name;
+    } else {
+      throw Exception('error: domain should be 1 parameter');
```
Comment (KNawm) [inline lib/src/rbac/default_role_manager.dart:64]: I would prefer an `ArgumentError` here.

File: lib/src/rbac/default_role_manager.dart [hunk 2]
```diff
@@ -12,46 +12,114 @@
...
   @override
-  void addLink(String name1, String name2, [String? domain]) {
-    if (domain != null) {
-      name1 = domain + '::' + name1;
-      name2 = domain + '::' + name2;
+  void addLink(String name1, String name2, [List<String> domain = const []]) {
+    if (domain.isEmpty) {
+      domain = [DEFAULT_DOMAIN];
+    } else if (domain.length > 1) {
+      throw Exception('error: domain should be 1 parameter');
```
Comment (KNawm) [inline lib/src/rbac/default_role_manager.dart:116]: Same as above.

File: lib/src/rbac/default_role_manager.dart [hunk 3]
```diff
@@ -60,85 +128,111 @@ class DefaultRoleManager implements RoleManager {
...
-  void deleteLink(String name1, String name2, [String? domain]) {
-    if (domain != null) {
-      name1 = domain + '::' + name1;
-      name2 = domain + '::' + name2;
+  void deleteLink(String name1, String name2,
+      [List<String> domain = const []]) {
+    if (domain.isEmpty) {
+      domain = [DEFAULT_DOMAIN];
+    } else if (domain.length > 1) {
+      throw Exception('error: domain should be 1 parameter');
```
Comment (KNawm) [inline lib/src/rbac/default_role_manager.dart:136]: Same as above.

File: lib/src/rbac/default_role_manager.dart [hunk 4]
```diff
@@ -60,85 +128,111 @@ class DefaultRoleManager implements RoleManager {
...
     }
 
-    if (!hasRole(name1) || !hasRole(name2)) {
+    final allRoles = loadOrDefault(allDomains, domain[0], DomainRoles());
+
+    if (!allRoles.roles.containsKey(name1) ||
+        !allRoles.roles.containsKey(name2)) {
       throw ArgumentError('error: name1 or name2 does not exist');
     }
```
Comment (KNawm) [inline lib/src/rbac/default_role_manager.dart:144]: Maybe check each parameter on its own so we can let know the developer which parameter, if any, doesn't exist.

File: lib/src/rbac/default_role_manager.dart (hunk 1 above)
Comment (KNawm) [reply lib/src/rbac/default_role_manager.dart:64]: Also, the error message could be clearer in this case.

File: lib/src/rbac/default_role_manager.dart [hunk 5]
```diff
@@ -60,85 +128,111 @@ class DefaultRoleManager implements RoleManager {
...
   @override
-  bool hasLink(String name1, String name2, [String? domain]) {
-    if (domain != null) {
-      name1 = domain + '::' + name1;
-      name2 = domain + '::' + name2;
+  bool hasLink(String name1, String name2, [List<String> domain = const []]) {
+    if (domain.isEmpty) {
+      domain = [DEFAULT_DOMAIN];
+    } else if (domain.length > 1) {
+      throw Exception('error: domain should be 1 parameter');
```
Comment (KNawm) [inline lib/src/rbac/default_role_manager.dart:158]: I would prefer an `ArgumentError` here.

File: lib/src/rbac/default_role_manager.dart [hunk 6]
```diff
@@ -60,85 +128,111 @@ class DefaultRoleManager implements RoleManager {
...
   /// [domain] is a prefix to the roles.
   @override
-  List<String> getRoles(String name, [String? domain]) {
-    if (domain != null) {
-      name = domain + '::' + name;
+  List<String> getRoles(String name, [List<String> domain = const []]) {
+    if (domain.isEmpty) {
+      domain = [DEFAULT_DOMAIN];
+    } else if (domain.length > 1) {
+      throw Exception('error: domain should be 1 parameter');
```
Comment (KNawm) [inline lib/src/rbac/default_role_manager.dart:188]: Same as above.

File: lib/src/rbac/default_role_manager.dart [hunk 7]
```diff
@@ -60,85 +128,111 @@ class DefaultRoleManager implements RoleManager {
...
   /// Returns the users that inherits a subject.
   @override
-  List<String> getUsers(String name) {
-    if (!hasRole(name)) {
-      throw ArgumentError('error: name does not exist');
+  List<String> getUsers(String name, [List<String> domain = const []]) {
+    if (domain.isEmpty) {
+      domain = [DEFAULT_DOMAIN];
+    } else if (domain.length > 1) {
+      throw Exception('error: domain should be 1 parameter');
```
Comment (KNawm) [inline lib/src/rbac/default_role_manager.dart:212]: Same as above.

**Review by KNawm — CHANGES_REQUESTED** [review]

File: lib/src/rbac/default_role_manager.dart (hunk 1 above)
Comment (MrUnfunny) [reply lib/src/rbac/default_role_manager.dart]: I've made all the required changes. Please verify.

**Review by KNawm — APPROVED** [review]

**Pull request merged**